url9512d8fa89_bdkosher_card_games_tgz-pJ8-org_washcom_cardgames_battleroyale_BattleRoyaleGameSimulatorJ8/src/org/washcom/cardgames/battleroyale/ThreeBattlerAssessor.java

```diff
@@ -0,0 +1,120 @@
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+import static org.washcom.cardgames.battleroyale.Rules.*;
+import static org.washcom.cardgames.core.Denomination.ACE;
+import static org.washcom.cardgames.core.Denomination.TWO;
+
+/**
+ * A BattleAssessor for when there's exactly three battlers in the battle. If
+ * not, the methods will throw IllegalArgumentExceptions.
+ *
+ * @author Joe
+ */
+public class ThreeBattlerAssessor implements BattleAssessor {
+
+    private static final Logger log = Logger.getLogger(ThreeBattlerAssessor.class.toString());
+
+    @Override
+    public BattleCard pickWinner(Battle battle) {
+        checkBattlers(battle.getBattlers());
+        List<BattleCard> cards = battle.getBattleCards();
+        BattleCard one = cards.get(0);
+        BattleCard two = cards.get(1);
+        BattleCard three = cards.get(2);
+
+        /* J8 combo means that the players immediately swap decks. Multiple
+         * swaps can occur. XXX - verify the correctness of this
+         */
+        if (isJackEightCombo(one, two)) {
+            battle.getGame().swapHands(one.getPlayedBy(), two.getPlayedBy());
+        }
+        if (isJackEightCombo(one, three)) {
+            battle.getGame().swapHands(one.getPlayedBy(), three.getPlayedBy());
+        }
+        if (isJackEightCombo(two, three)) {
+            battle.getGame().swapHands(two.getPlayedBy(), three.getPlayedBy());
+        }
+
+        /*
+         * In a Battle Royale scenario, no one wins
+         */
+        if (isBattleRoyale(one, two) || isBattleRoyale(one, three) || isBattleRoyale(two, three)) {
+            return null;
+        }
+
+        /*
+         * Determine value difference between three players. If any combination is
+         * under the battle threshold, there is no winner.
+         */
+        int oneTwoDiff = one.computeValueDifference(two);
+        int oneThreeDiff = one.computeValueDifference(three);
+        int twoThreeDiff = two.computeValueDifference(three);
+        if (isDifferentialUnderBattleThreshold(oneTwoDiff)
+                || isDifferentialUnderBattleThreshold(oneThreeDiff)
+                || isDifferentialUnderBattleThreshold(twoThreeDiff)) {
+            return null;
+        }
+        BattleCard highCard = highCard(one, two, three);
+        /*
+         * If an Ace is the high card, check for a two. In an Ace-Two combo, the Two wins.
+         * It's one of those quirky rules.
+         */
+        if (highCard.getCard().getDenomination() == ACE) {
+            if (one.getCard().getDenomination() == TWO) {
+                return one;
+            } else if (two.getCard().getDenomination() == TWO) {
+                return two;
+            } else if (three.getCard().getDenomination() == TWO) {
+                return three;
+            }
+        }
+        return highCard;
+    }
+
+    /* FIXME */
+    @Override
+    public Map<Player, Integer> determineFees(Battle battle) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 89-100, lines 102-113 into separate methods.